src/util/newconfig/ConfigDefinition.cpp

```diff
@@ -0,0 +1,209 @@
+     {"cache.num_cursors_from_diff", ConfigValue{ConfigType::Integer}.defaultValue(0)},
+     {"cache.num_cursors_from_account", ConfigValue{ConfigType::Integer}.defaultValue(0)},
+     {"cache.page_fetch_size", ConfigValue{ConfigType::Integer}.defaultValue(512)},
+     {"cache.load", ConfigValue{ConfigType::String}.defaultValue("async")},
+     {"log_channels.[].channel", Array{ConfigValue{ConfigType::String}.optional()}},
+     {"log_channels.[].log_level", Array{ConfigValue{ConfigType::String}.optional()}},
+     {"log_level", ConfigValue{ConfigType::String}.defaultValue("info")},
+     {"log_format",
+      ConfigValue{ConfigType::String}.defaultValue(
+          R"(%TimeStamp% (%SourceLocation%) [%ThreadID%] %Channel%:%Severity% %Message%)"
+      )},
+     {"log_to_console", ConfigValue{ConfigType::Boolean}.defaultValue(false)},
+     {"log_directory", ConfigValue{ConfigType::String}.optional()},
+     {"log_rotation_size", ConfigValue{ConfigType::Integer}.defaultValue(2048)},
+     {"log_directory_max_size", ConfigValue{ConfigType::Integer}.defaultValue(50 * 1024)},
+     {"log_rotation_hour_interval", ConfigValue{ConfigType::Integer}.defaultValue(12)},
+     {"log_tag_style", ConfigValue{ConfigType::String}.defaultValue("uint")},
+     {"extractor_threads", ConfigValue{ConfigType::Integer}.defaultValue(2u)},
+     {"read_only", ConfigValue{ConfigType::Boolean}.defaultValue(false)},
+     {"txn_threshold", ConfigValue{ConfigType::Integer}.defaultValue(0)},
+     {"start_sequence", ConfigValue{ConfigType::String}.optional()},
+     {"finish_sequence", ConfigValue{ConfigType::String}.optional()},
+     {"ssl_cert_file", ConfigValue{ConfigType::String}.optional()},
+     {"ssl_key_file", ConfigValue{ConfigType::String}.optional()},
+     {"api_version.min", ConfigValue{ConfigType::Integer}},
+     {"api_version.max", ConfigValue{ConfigType::Integer}}}
+};
+
+ClioConfigDefinition::ClioConfigDefinition(std::initializer_list<KeyValuePair> pair)
+{
+    for (auto const& p : pair) {
+        if (p.first.contains("[]"))
+            ASSERT(std::holds_alternative<Array>(p.second), "Value must be array if key has \"[]\"");
+        map_.insert(p);
+    }
+}
+
+ObjectView
+ClioConfigDefinition::getObject(std::string_view prefix, std::optional<std::size_t> idx) const
+{
+    auto const prefixWithDot = std::string(prefix) + ".";
+    for (auto const& [mapKey, mapVal] : map_) {
+        auto const hasPrefix = mapKey.starts_with(prefixWithDot);
+        if (idx.has_value() && hasPrefix && std::holds_alternative<Array>(mapVal)) {
+            ASSERT(std::get<Array>(mapVal).size() > idx.value(), "Index provided is out of scope");
+            // we want to support getObject("array") and getObject("array.[]"), so we check if "[]" exists
+            if (!prefix.contains("[]"))
+                return ObjectView{prefixWithDot + "[]", idx.value(), *this};
+            return ObjectView{prefix, idx.value(), *this};
```

Comment:
How is this possible when prefix is not an array, but `idx` has value?

Reply:
In this case, prefix is always an array, (i.e in Clio Config should be defined as array.[].something), but we want to support getObject("array") and getObject("array.[]"), so I did the explicit check to see if "[]" exists, as prefix passed into objectView should contain "array.[]" not "array"